File: Sleepscape/Drawing/DrawingEngine.swift
import Foundation
import Metal
import simd

/// Core drawing engine: manages strokes, ripples, ambient blooms,
/// sparkle particles, and undo/redo history.
class DrawingEngine {
    private(set) var strokes: [Stroke] = []
    private(set) var currentStroke: Stroke?
    private(set) var predictedPoints: [CGPoint] = []

    // Undo/Redo stack
    private var undoneStrokes: [Stroke] = []
    var canUndo: Bool { !strokes.isEmpty }
    var canRedo: Bool { !undoneStrokes.isEmpty }

    // Smoothing state (exponential moving average — more fluid)
    private var sx: Float = 0
    private var sy: Float = 0
    private var isFirstPoint = true
    private let smoothFactor: Float = 0.08       // finger: lower = smoother, more meditative
    private let pencilSmoothFactor: Float = 0.35  // pencil: light easing, preserves accuracy
    private let slowInkFactor: Float = 0.06       // slow ink: very heavy smoothing
    var slowInkEnabled: Bool = false

    // Cumulative distance tracking
    private var lastX: Float = 0
    private var lastY: Float = 0
    private var cumulativeDistance: Float = 0

    // MARK: - Ripple state

    struct Ripple {
        var center: SIMD2<Float>
        var radius: Float
        var maxRadius: Float
        var alpha: Float
        var speed: Float
        var color: SIMD4<Float>
        var rings: Int
    }
    private(set) var ripples: [Ripple] = []
    private var lastRippleTime: CFTimeInterval = 0

    // MARK: - Ambient Bloom state

    struct AmbientBloom {
        var center: SIMD2<Float>
        var radius: Float
        var maxRadius: Float
        var alpha: Float
        var targetAlpha: Float
        var color: SIMD4<Float>
        var phase: BloomPhase
        var phaseTimer: Float  // frames remaining in current phase
    }
    enum BloomPhase { case fadeIn, hold, expand, fadeOut }

    private(set) var ambientBlooms: [AmbientBloom] = []
    private var lastBloomSpawnTime: CFTimeInterval = 0
    var bloomsEnabled: Bool = true
    var bloomSpawnRate: Float = 0.5   // 0→1
    var bloomIntensity: Float = 0.6   // 0→1

    // MARK: - Sparkle particle state

    struct Sparkle {
        var position: SIMD2<Float>
        var alpha: Float
        var size: Float
        var color: SIMD4<Float>
        var life: Float   // frames remaining
    }
    private(set) var sparkles: [Sparkle] = []

    // MARK: - Touch handling

    func beginStroke(at point: CGPoint, pressure: Float, altitude: Float,
                     color: SIMD4<Float>, brushSize: Float, mode: DrawMode,
                     lineStyle: LineStyle = .neon) {
        isFirstPoint = true
        sx = Float(point.x)
        sy = Float(point.y)
        lastX = sx
        lastY = sy
        cumulativeDistance = 0

        var stroke = Stroke(
            colorR: color.x, colorG: color.y, colorB: color.z, colorA: color.w,
            brushSize: brushSize, mode: mode, lineStyle: lineStyle
        )
        stroke.points.append(StrokePoint(
            x: Float(point.x), y: Float(point.y),
            pressure: pressure, altitude: altitude, cumulDist: 0
        ))
        currentStroke = stroke

        // Clear redo stack on new stroke
        undoneStrokes.removeAll()
    }

    func addPoint(_ point: CGPoint, pressure: Float, altitude: Float,
                  isPencil: Bool, canvasSize: CGSize? = nil) {
        guard currentStroke != nil else { return }

        var px = Float(point.x)
        var py = Float(point.y)

        // Apply smoothing: slow ink overrides, otherwise finger/pencil defaults
        let factor = slowInkEnabled ? slowInkFactor : (isPencil ? pencilSmoothFactor : smoothFactor)
        if isFirstPoint {
            sx = px
            sy = py
            isFirstPoint = false
        } else {
            sx += (px - sx) * factor
            sy += (py - sy) * factor
            px = sx
            py = sy
        }

        // Update cumulative distance
        let dx = px - lastX
        let dy = py - lastY
        let segLen = sqrt(dx * dx + dy * dy)
        cumulativeDistance += segLen
        lastX = px
        lastY = py

        currentStroke?.points.append(StrokePoint(
            x: px, y: py,
            pressure: pressure, altitude: altitude,
            cumulDist: cumulativeDistance
        ))

        // Spawn sparkle particles along pencil strokes
        if isPencil && segLen > 2.0 && sparklesEnabled {
            spawnSparkle(at: SIMD2<Float>(px, py),
                        color: SIMD4<Float>(currentStroke!.colorR,
                                            currentStroke!.colorG,
                                            currentStroke!.colorB, 1.0))
        }
    }

    func setPredicted(_ points: [CGPoint]) {
        predictedPoints = points
    }

    var pathSmoothingEnabled: Bool = false
    var sparklesEnabled: Bool = true

    func endStroke() {
        if var stroke = currentStroke, !stroke.points.isEmpty {
            if pathSmoothingEnabled && stroke.points.count >= 3 {
                stroke.points = chaikinSmooth(stroke.points, iterations: 5)
                // Cap at 500 points to prevent memory issues on long strokes
                if stroke.points.count > 500 {
                    stroke.points = downsample(stroke.points, to: 500)
                }
            }
            strokes.append(stroke)
        }
        currentStroke = nil
        predictedPoints = []
    }

    /// Uniform stride downsample — keeps first and last, evenly samples the rest.
    private func downsample(_ points: [StrokePoint], to count: Int) -> [StrokePoint] {
        guard points.count > count, count >= 2 else { return points }
        var result = [StrokePoint]()
        let stride = Float(points.count - 1) / Float(count - 1)
        for i in 0..<count {
            let idx = min(Int(Float(i) * stride), points.count - 1)
            result.append(points[idx])
        }
        return result
    }

    // MARK: - Chaikin curve refinement

    /// Runs Chaikin's corner-cutting algorithm to produce smooth bezier-like curves.
    private func chaikinSmooth(_ points: [StrokePoint], iterations: Int) -> [StrokePoint] {
        var pts = points
        for _ in 0..<iterations {
            guard pts.count >= 2 else { return pts }
            var newPts = [StrokePoint]()
            newPts.append(pts[0])  // keep first point
            for i in 0..<(pts.count - 1) {
                let a = pts[i], b = pts[i + 1]
                // Q = 75% A + 25% B
                let q = StrokePoint(
                    x: a.x * 0.75 + b.x * 0.25,
                    y: a.y * 0.75 + b.y * 0.25,
                    pressure: a.pressure * 0.75 + b.pressure * 0.25,
                    altitude: a.altitude * 0.75 + b.altitude * 0.25,
                    cumulDist: 0
                )
                // R = 25% A + 75% B
                let r = StrokePoint(
                    x: a.x * 0.25 + b.x * 0.75,
                    y: a.y * 0.25 + b.y * 0.75,
                    pressure: a.pressure * 0.25 + b.pressure * 0.75,
                    altitude: a.altitude * 0.25 + b.altitude * 0.75,
                    cumulDist: 0
                )
                newPts.append(q)
                newPts.append(r)
            }
            newPts.append(pts.last!)  // keep last point
            pts = newPts
        }
        // Recalculate cumulative distance
        var dist: Float = 0
        pts[0].cumulDist = 0
        for i in 1..<pts.count {
            let dx = pts[i].x - pts[i - 1].x
            let dy = pts[i].y - pts[i - 1].y
            dist += sqrt(dx * dx + dy * dy)
            pts[i].cumulDist = dist
        }
        return pts
    }

    // MARK: - Undo / Redo

    func undo() -> Bool {
        guard let last = strokes.popLast() else { return false }
        undoneStrokes.append(last)
        return true
    }

    func redo() -> Bool {
        guard let restored = undoneStrokes.popLast() else { return false }
        strokes.append(restored)
        return true
    }

    // MARK: - Ripples

    func spawnRipples(at centers: [SIMD2<Float>], color: SIMD4<Float>, time: CFTimeInterval, reach: Float = 0.5) {
        guard time - lastRippleTime >= 0.700 else { return }
        lastRippleTime = time

        let speed: Float = 0.30 + Float.random(in: 0...0.12)
        let baseMax: Float = 120 + reach * 280  // reach 0→120, reach 1→400
        let maxR: Float = baseMax + Float.random(in: 0...30)

        for center in centers {
            let ripple = Ripple(
                center: center,
                radius: 30,
                maxRadius: maxR,
                alpha: 0.85,
                speed: speed,
                color: color,
                rings: 3
            )
            ripples.append(ripple)
        }
    }

    func updateRipples() {
        for i in ripples.indices.reversed() {
            ripples[i].radius += ripples[i].speed
            let progress = ripples[i].radius / ripples[i].maxRadius
            let decay: Float = 1.0 - (0.002 + 0.025 * progress * progress)
            ripples[i].alpha *= decay
            if ripples[i].alpha < 0.01 || ripples[i].radius > ripples[i].maxRadius {
                ripples.remove(at: i)
            }
        }
    }

    // MARK: - Ambient Blooms

    func updateAmbientBlooms(canvasSize: CGSize, inkColor: SIMD4<Float>, time: CFTimeInterval) {
        guard bloomsEnabled else {
            ambientBlooms.removeAll()
            return
        }

        // Spawn new blooms at random intervals — small, scattered, delicate
        let spawnInterval = Double(3.0 - bloomSpawnRate * 2.5)
        if time - lastBloomSpawnTime >= spawnInterval && ambientBlooms.count < 8 {
            lastBloomSpawnTime = time

            // Scatter across full canvas
            let cx = Float.random(in: 80...Float(canvasSize.width) - 80)
            let cy = Float.random(in: 80...Float(canvasSize.height) - 80)
            let maxR = Float.random(in: 60...180)  // generous expansion range

            ambientBlooms.append(AmbientBloom(
                center: SIMD2<Float>(cx, cy),
                radius: 3,
                maxRadius: maxR,
                alpha: 0,
                targetAlpha: bloomIntensity * Float.random(in: 0.20...0.40),
                color: inkColor,
                phase: .fadeIn,
                phaseTimer: Float.random(in: 80...140)  // slow fade-in (~1s)
            ))
        }

        // Update: slow firework — fade in briefly, then continuously expand + dissolve
        for i in ambientBlooms.indices.reversed() {
            ambientBlooms[i].phaseTimer -= 1

            let progress = ambientBlooms[i].radius / ambientBlooms[i].maxRadius

            // Very slow expansion — decelerates as it grows (like real dissipation)
            let expandSpeed: Float = 0.08 + (1.0 - progress) * 0.12
            ambientBlooms[i].radius += expandSpeed

            switch ambientBlooms[i].phase {
            case .fadeIn:
                // Fade-in over ~0.5 second
                let fadeRate = ambientBlooms[i].targetAlpha / 60.0
                ambientBlooms[i].alpha = min(ambientBlooms[i].alpha + fadeRate, ambientBlooms[i].targetAlpha)
                if ambientBlooms[i].phaseTimer <= 0 {
                    // Skip hold — go straight to dissolve
                    ambientBlooms[i].phase = .fadeOut
                    ambientBlooms[i].phaseTimer = 999
                }

            case .hold:
                ambientBlooms[i].phase = .fadeOut

            case .expand:
                ambientBlooms[i].phase = .fadeOut

            case .fadeOut:
                // Very gradual continuous fade — slower at start, faster as it expands
                // Like a firework ember slowly losing its glow
                let fadeFactor: Float = 1.0 - (0.002 + 0.008 * progress * progress)
                ambientBlooms[i].alpha *= fadeFactor
            }

            if ambientBlooms[i].alpha < 0.002 || ambientBlooms[i].radius > ambientBlooms[i].maxRadius {
                ambientBlooms.remove(at: i)
            }
        }
    }

    // MARK: - Sparkle particles

    private func spawnSparkle(at position: SIMD2<Float>, color: SIMD4<Float>) {
        guard sparkles.count < 200 else { return }

        // Spawn 1-3 sparkles near the point
        let count = Int.random(in: 1...3)
        for _ in 0..<count {
            let offset = SIMD2<Float>(Float.random(in: -8...8), Float.random(in: -8...8))
            sparkles.append(Sparkle(
                position: position + offset,
                alpha: Float.random(in: 0.4...0.9),
                size: Float.random(in: 2...5),
                color: color,
                life: Float.random(in: 20...50) // frames
            ))
        }
    }

    func updateSparkles() {
        for i in sparkles.indices.reversed() {
            sparkles[i].life -= 1
            sparkles[i].alpha *= 0.95  // gentle fade
            sparkles[i].size *= 0.98   // shrink slightly
            if sparkles[i].life <= 0 || sparkles[i].alpha < 0.01 {
                sparkles.remove(at: i)
            }
        }
    }

    // MARK: - Clear

    func clearAll() {
        strokes.removeAll()
        undoneStrokes.removeAll()
        currentStroke = nil
        predictedPoints = []
        ripples.removeAll()
        ambientBlooms.removeAll()
        sparkles.removeAll()
    }
}

File: web/src/drawing/DrawingEngine.ts
import type { AppState } from '../state/AppState';
import type { StrokeRenderer } from '../rendering/StrokeRenderer';
import type { EllipseRenderer } from '../rendering/EllipseRenderer';
import type { Stroke, StrokePoint, EllipseShape } from '../state/types';
import type { Ripple, Sparkle, AmbientBloom } from '../rendering/ParticleRenderer';
import { DrawMode } from '../state/types';
import { generateTransforms, transformCount } from '../rendering/SymmetryTransform';
import { chaikinSmooth, uid } from '../utils/math';

type RenderEntry =
  | { type: 'stroke'; stroke: Stroke }
  | { type: 'ellipse'; ellipse: EllipseShape };

type ActionEntry =
  | { type: 'stroke'; stroke: Stroke }
  | { type: 'ellipseGesture'; ellipses: EllipseShape[] };

interface ActiveEllipseGesture {
  center: [number, number];
  color: [number, number, number, number];
  lineWidth: number;
  mode: DrawMode;
  symmetry: number;
  lineStyle: Stroke['lineStyle'];
  ellipses: EllipseShape[];
}

/**
 * Manages drawing (strokes + ellipses), undo/redo, and particle systems.
 */
export class DrawingEngine {
  private state: AppState;
  private strokeRenderer: StrokeRenderer;
  private ellipseRenderer: EllipseRenderer;
  private getCanvasSize: () => { width: number; height: number };

  // Drawing data
  strokes: Stroke[] = [];
  ellipses: EllipseShape[] = [];
  private renderEntries: RenderEntry[] = [];
  private actions: ActionEntry[] = [];
  private undoneActions: ActionEntry[] = [];
  private activeStroke: Stroke | null = null;
  private activeEllipse: ActiveEllipseGesture | null = null;

  // Particles
  ripples: Ripple[] = [];
  sparkles: Sparkle[] = [];
  ambientBlooms: AmbientBloom[] = [];

  private lastRippleTime = 0;
  private lastBloomSpawnTime = 0;

  constructor(
    state: AppState,
    strokeRenderer: StrokeRenderer,
    ellipseRenderer: EllipseRenderer,
    getCanvasSize: () => { width: number; height: number },
  ) {
    this.state = state;
    this.strokeRenderer = strokeRenderer;
    this.ellipseRenderer = ellipseRenderer;
    this.getCanvasSize = getCanvasSize;
  }

  beginStroke(x: number, y: number, pressure: number, altitude: number): void {
    if (this.state.drawMode === DrawMode.Ellipse) {
      this.activeStroke = null;
      this.activeEllipse = {
        center: [x, y],
        color: this.state.currentInkRGBA,
        lineWidth: this.state.brushSize,
        mode: this.state.drawMode,
        symmetry: this.state.symmetry,
        lineStyle: this.state.lineStyle,
        ellipses: [],
      };
    } else {
      const point: StrokePoint = { x, y, pressure, altitude, cumulDist: 0 };
      this.activeEllipse = null;
      this.activeStroke = {
        id: uid(),
        points: [point],
        color: this.state.currentInkRGBA,
        brushSize: this.state.brushSize,
        mode: this.state.drawMode,
        lineStyle: this.state.lineStyle,
        symmetry: this.state.symmetry,
      };
    }

    // Spawn initial ripple
    if (this.state.ripplesEnabled) {
      this.spawnRipples(x, y);
    }
  }

  addPoint(x: number, y: number, pressure: number, altitude: number, isPencil = false): void {
    if (this.activeEllipse) {
      this.addEllipsePoint(x, y);

      if (this.state.ripplesEnabled) {
        this.spawnRipples(x, y);
      }
      return;
    }

    if (!this.activeStroke) return;

    const points = this.activeStroke.points;
    const prev = points[points.length - 1];

    const dx = x - prev.x;
    const dy = y - prev.y;
    const dist = Math.sqrt(dx * dx + dy * dy);

    // Skip if too close
    if (dist < 1) return;

    const point: StrokePoint = {
      x, y, pressure, altitude,
      cumulDist: prev.cumulDist + dist,
    };

    points.push(point);

    // Render the new segment incrementally
    if (points.length >= 2) {
      this.renderIncrementalSegment();
    }

    // Spawn ripples along stroke
    if (this.state.ripplesEnabled) {
      this.spawnRipples(x, y);
    }

    // Spawn sparkles along pencil strokes
    if (isPencil && dist > 2 && this.state.sparklesEnabled) {
      this.spawnSparkles(x, y);
    }
  }

  endStroke(): void {
    if (this.activeEllipse) {
      this.endEllipseGesture();
      return;
    }
    if (!this.activeStroke) return;

    const stroke = this.activeStroke;
    this.activeStroke = null;

    // Apply path smoothing if enabled
    let didSmooth = false;
    if (this.state.pathSmoothingEnabled && stroke.points.length > 2) {
      stroke.points = chaikinSmooth(stroke.points);
      didSmooth = true;
    }

    this.strokes.push(stroke);
    this.renderEntries.push({ type: 'stroke', stroke });
    this.actions.push({ type: 'stroke', stroke });
    this.undoneActions = [];

    // Re-render only when smoothing changed points already drawn incrementally.
    if (didSmooth) {
      this.reRenderAll();
    }

    this.updateUndoRedoState();
  }

  undo(): void {
    if (this.actions.length === 0) return;

    const action = this.actions.pop()!;
    this.undoneActions.push(action);

    if (action.type === 'stroke') {
      this.strokes.pop();
      this.renderEntries.pop();
    } else {
      const removeCount = action.ellipses.length;
      if (removeCount > 0) {
        this.ellipses.splice(Math.max(0, this.ellipses.length - removeCount), removeCount);
        this.renderEntries.splice(Math.max(0, this.renderEntries.length - removeCount), removeCount);
      }
    }

    this.reRenderAll();
    this.updateUndoRedoState();
  }

  redo(): void {
    if (this.undoneActions.length === 0) return;

    const action = this.undoneActions.pop()!;
    this.actions.push(action);

    if (action.type === 'stroke') {
      this.strokes.push(action.stroke);
      this.renderEntries.push({ type: 'stroke', stroke: action.stroke });
      this.strokeRenderer.renderStroke(action.stroke, this.state.glowIntensity);
    } else {
      for (const ellipse of action.ellipses) {
        this.ellipses.push(ellipse);
        this.renderEntries.push({ type: 'ellipse', ellipse });
        this.ellipseRenderer.renderEllipse(ellipse, this.state.glowIntensity);
      }
    }

    this.updateUndoRedoState();
  }

  reRenderAll(): void {
    // Clear the persistent drawing surface.
    this.strokeRenderer.reRenderAll([], this.state.glowIntensity);
    for (const entry of this.renderEntries) {
      if (entry.type === 'stroke') {
        this.strokeRenderer.renderStroke(entry.stroke, this.state.glowIntensity);
      } else {
        this.ellipseRenderer.renderEllipse(entry.ellipse, this.state.glowIntensity);
      }
    }
  }

  clear(): void {
    this.strokes = [];
    this.ellipses = [];
    this.renderEntries = [];
    this.actions = [];
    this.undoneActions = [];
    this.activeStroke = null;
    this.activeEllipse = null;
    this.ripples = [];
    this.sparkles = [];
    this.ambientBlooms = [];
    this.state.canUndo = false;
    this.state.canRedo = false;
  }

  // ── Particle updates (called each frame) ──

  updateParticles(canvasWidth: number, canvasHeight: number, time: number): void {
    this.updateRipples();
    this.updateSparkles();
    this.updateAmbientBlooms(canvasWidth, canvasHeight, time);
  }

  private updateRipples(): void {
    for (let i = this.ripples.length - 1; i >= 0; i--) {
      const r = this.ripples[i];
      r.radius += r.speed;
      const progress = r.radius / r.maxRadius;
      const decay = 1.0 - (0.002 + 0.025 * progress * progress);
      r.alpha *= decay;
      if (r.alpha < 0.01 || r.radius > r.maxRadius) {
        this.ripples.splice(i, 1);
      }
    }
  }

  private updateSparkles(): void {
    for (let i = this.sparkles.length - 1; i >= 0; i--) {
      const s = this.sparkles[i];
      s.life -= 1;
      s.alpha *= 0.95;
      s.size *= 0.98;
      if (s.life <= 0 || s.alpha < 0.01) {
        this.sparkles.splice(i, 1);
      }
    }
  }

  private updateAmbientBlooms(canvasWidth: number, canvasHeight: number, time: number): void {
    if (!this.state.bloomsEnabled) {
      this.ambientBlooms.length = 0;
      return;
    }

    // Spawn new blooms at random intervals
    const spawnInterval = 3.0 - this.state.bloomSpawnRate * 2.5;
    if (time - this.lastBloomSpawnTime >= spawnInterval && this.ambientBlooms.length < 8) {
      this.lastBloomSpawnTime = time;

      const [r, g, b] = [
        this.state.currentInkRGBA[0],
        this.state.currentInkRGBA[1],
        this.state.currentInkRGBA[2],
      ];

      this.ambientBlooms.push({
        centerX: 80 + Math.random() * (canvasWidth - 160),
        centerY: 80 + Math.random() * (canvasHeight - 160),
        radius: 3,
        maxRadius: 60 + Math.random() * 120,
        alpha: 0,
        targetAlpha: this.state.bloomIntensity * (0.20 + Math.random() * 0.20),
        colorR: r, colorG: g, colorB: b,
        phase: 'fadeIn',
        phaseTimer: 80 + Math.random() * 60,
      });
    }

    // Update existing blooms
    for (let i = this.ambientBlooms.length - 1; i >= 0; i--) {
      const b = this.ambientBlooms[i];
      b.phaseTimer -= 1;

      const progress = b.radius / b.maxRadius;
      const expandSpeed = 0.08 + (1.0 - progress) * 0.12;
      b.radius += expandSpeed;

      if (b.phase === 'fadeIn') {
        const fadeRate = b.targetAlpha / 60.0;
        b.alpha = Math.min(b.alpha + fadeRate, b.targetAlpha);
        if (b.phaseTimer <= 0) {
          b.phase = 'fadeOut';
          b.phaseTimer = 999;
        }
      } else {
        // fadeOut: gradual continuous fade
        const fadeFactor = 1.0 - (0.002 + 0.008 * progress * progress);
        b.alpha *= fadeFactor;
      }

      if (b.alpha < 0.002 || b.radius > b.maxRadius) {
        this.ambientBlooms.splice(i, 1);
      }
    }
  }

  // ── Particle spawning ──

  private spawnRipples(x: number, y: number): void {
    const now = performance.now() / 1000;
    if (now - this.lastRippleTime < 2.0) return;
    this.lastRippleTime = now;

    const [cr, cg, cb] = [
      this.state.currentInkRGBA[0],
      this.state.currentInkRGBA[1],
      this.state.currentInkRGBA[2],
    ];
    const reach = (this.state as any).rippleReach ?? 0.5;
    const baseMax = 120 + reach * 280;

    // Spawn at all symmetry-mirrored positions.
    const mode = this.activeStroke?.mode ?? this.activeEllipse?.mode ?? this.state.drawMode;
    const sym = this.activeStroke?.symmetry ?? this.activeEllipse?.symmetry ?? this.state.symmetry;
    const { width: canvasW, height: canvasH } = this.getCanvasSize();
    const transforms = generateTransforms(mode, sym, canvasW, canvasH);
    const count = transformCount(mode, sym);

    for (let t = 0; t < count; t++) {
      const o = t * 9;
      // Apply 3x3 transform (column-major) to point.
      const tx = transforms[o + 0] * x + transforms[o + 3] * y + transforms[o + 6];
      const ty = transforms[o + 1] * x + transforms[o + 4] * y + transforms[o + 7];

      // Jitter position for organic feel
      const jx = (Math.random() - 0.5) * 20;
      const jy = (Math.random() - 0.5) * 20;
      const cx = tx + jx;
      const cy = ty + jy;

      // Spawn 2 concentric rings like a stone in water
      const baseSpeed = 0.18 + Math.random() * 0.10;
      const ringStarts = [3, 22];
      const ringAlphas = [0.40, 0.20];
      const ringSpeeds = [baseSpeed, baseSpeed * 0.85];
      const maxR = baseMax + (Math.random() - 0.5) * 50;

      for (let r = 0; r < 2; r++) {
        this.ripples.push({
          centerX: cx, centerY: cy,
          radius: ringStarts[r], maxRadius: maxR,
          alpha: ringAlphas[r], speed: ringSpeeds[r],
          colorR: cr, colorG: cg, colorB: cb,
          rings: 3,
        });
      }
    }
  }

  private spawnSparkles(x: number, y: number): void {
    if (this.sparkles.length >= 200) return;

    const [cr, cg, cb] = [
      this.state.currentInkRGBA[0],
      this.state.currentInkRGBA[1],
      this.state.currentInkRGBA[2],
    ];

    const count = 1 + Math.floor(Math.random() * 3);
    for (let i = 0; i < count; i++) {
      this.sparkles.push({
        x: x + (Math.random() - 0.5) * 16,
        y: y + (Math.random() - 0.5) * 16,
        alpha: 0.4 + Math.random() * 0.5,
        size: 2 + Math.random() * 3,
        colorR: cr, colorG: cg, colorB: cb,
        life: 20 + Math.random() * 30,
      });
    }
  }

  // ── Incremental rendering ──

  private renderIncrementalSegment(): void {
    if (!this.activeStroke || this.activeStroke.points.length < 2) return;

    const points = this.activeStroke.points;
    const miniStroke: Stroke = {
      ...this.activeStroke,
      points: [points[points.length - 2], points[points.length - 1]],
    };

    this.strokeRenderer.renderStroke(miniStroke, this.state.glowIntensity);
  }

  private addEllipsePoint(x: number, y: number): void {
    if (!this.activeEllipse) return;

    const [cx, cy] = this.activeEllipse.center;
    const rx = Math.abs(x - cx);
    const ry = Math.abs(y - cy);
    if (rx < 2 && ry < 2) return;

    const ellipse: EllipseShape = {
      center: this.activeEllipse.center,
      radii: [rx, ry],
      color: this.activeEllipse.color,
      lineWidth: this.activeEllipse.lineWidth,
      mode: this.activeEllipse.mode,
      symmetry: this.activeEllipse.symmetry,
      lineStyle: this.activeEllipse.lineStyle,
    };

    // Replace the preview ellipse (only keep the latest one during drag)
    this.activeEllipse.ellipses = [ellipse];

    // Re-render everything + the preview ellipse
    this.reRenderAll();
    this.ellipseRenderer.renderEllipse(ellipse, this.state.glowIntensity);
  }

  private endEllipseGesture(): void {
    if (!this.activeEllipse) return;

    const ellipses = this.activeEllipse.ellipses;
    this.activeEllipse = null;

    if (ellipses.length > 0) {
      this.ellipses.push(...ellipses);
      for (const ellipse of ellipses) {
        this.renderEntries.push({ type: 'ellipse', ellipse });
      }
      this.actions.push({ type: 'ellipseGesture', ellipses });
      this.undoneActions = [];
    }

    this.updateUndoRedoState();
  }

  private updateUndoRedoState(): void {
    this.state.canUndo = this.actions.length > 0;
    this.state.canRedo = this.undoneActions.length > 0;
  }
}
